engineering-plans/0002-ethereum-tracing-cache.md

```diff
@@ -0,0 +1,122 @@
+# PLAN-0002: Ethereum Tracing Cache
+
+<dl>
+  <dt>Author</dt>
+  <dd>Zachary Burns</dd>
+
+  <dt>Implements</dt>
+  <dd><a href="../rfcs/0002-ethereum-tracing-cache.md">RFC-0002 Ethereum Tracing Cache</a></dd>
+
+  <dt>Engineering Plan pull request</dt>
+  <dd><a href="https://github.com/graphprotocol/rfcs/pull/9">https://github.com/graphprotocol/rfcs/pull/9</a></dd>
+
+  <dt>Obsoletes (if applicable)</dt>
+  <dd>None</dd>
+
+  <dt>Date of submission</dt>
+  <dd>2019-12-20</dd>
+
+  <dt>Date of approval</dt>
+  <dd>YYYY-MM-DD</dd>
+
+  <dt>Approved by</dt>
+  <dd>First Person, Second Person</dd>
+</dl>
+
+## Summary
+
+Implements RFC-0002: Ethereum Tracing Cache
+
+## Implementation
+
+These changes happen within `ethereum_adapter.rs`, `store.rs` and `db_schema.rs`.
+
+### ServiceCache trait
+EthereumCallCache will be renamed to ServiceCache and have roughly this signature:
+```rust
+trait ServiceCache {
+    // Provided function
+    fn get<T: Deserialize>(
+        // Arguments to fetch
+        args: impl Hash,
+        // Perform the work if not found in the cache
+        f: impl Future<Output=Result<Vec<u8>, Error>>,
+        // Uniquely identify the cache and it's version number. This is essentially a namespace.
+        cache_id: &str
+    ) -> Box<dyn Future<Output=Result<T, Error=Error>>>;
+
+    // Required functions
+    fn try_store(key: [u8; 16], data: &[u8]) -> Box<dyn Future<Output=Result<(), Error>>>;
+    fn try_get(key: [u8; 16]) -> Box<dyn Future<Output=Result<Option<Vec<u8>, Error>>>>;
+}
+```
+
+This signature aims to fix 2 problems:
+  * Today, EthereumCallCache get_call is blocking, and used by code which returns a Future. This is [not recommended for async code](https://doc.rust-lang.org/stable/std/future/trait.Future.html#runtime-characteristics) and can cause problems for both runtimes and future combinators. The new signature allows for the database work to be migrated to use non-blocking apis, or to be spawned in a threadpool.
+  * This signature moves the `get or call then set` logic into the trait where it only need be written once, rather than everywhere it is used. This also allows for certain optimizations, like only calculating the hash of the arguments once.
+
+The Hash trait would be from [digest-hash](https://crates.io/crates/digest-hash/0.3.0) allowing tiny_keccak use for data structures, taking over the roll that `contract_call_id` fills today.
```

Comment:
> `std::Hash` doesn't allow that? (https://doc.rust-lang.org/std/hash/index.html)

No, and for good reason. Originally I thought a good way to approach this would be to write an adaptor allowing `std::Hash` to be used with `digest::Digest` generally. Then we could just do `#[derive(Hash)]` and life would be easy.

But.. the implementation of Hash only guarantees stability within a single process and is not for inter-process use. This applies not just to the `Hasher` trait but the `Hash` trait as well. One issue is that some data types like `usize` are not portable. Another issue is that some data types like `*const u8` _should_ provide an implementation for `std::Hash` but not be used with `digest::Digest` ever. Another issue is that the compiler is free to re-order the fields in `#[derive(Hash)` each time that it compiles. Another is that `#[derive(Hash)]` or some implementation is free to inject process specific random data, like the `ProcessId` on Windows, or just some randomly generated number on startup, or `std::any::TypeId` because none of those break the invariant that the hash will be stable for the given process.

If we found ourselves needing this a lot, we might contribute back a `#[derive(Hash)]` for `digest-hash`. But, I don't think we need it frequently enough for that to be a big win on time.

Reply:
Fun. Ok, makes complete sense now, thanks for the insights!